### PR TITLE
Utilize New Types

### PR DESCRIPTION
**Pull request by NachoToast**
Moderate changes to the API to use the new object shapes (introduced in https://github.com/UoA-Discords/shared/pull/1).

### CODE REVIEW

File: src/classes/AuthDiscordAPI/AuthDiscordAPI.ts [hunk 1]
```diff
@@ -53,15 +53,13 @@ export abstract class AuthDiscordAPI {
...
      * @throws Throws an error if the provided access token is invalid.
      */
-    public static async revokeToken(token: string): Promise<boolean> {
+    public static async revokeToken(token: string): Promise<void> {
         const body = this.makeRequestBody();
         body.set(`token`, token);
 
-        const { data } = await axios.post<boolean>(OAuth2Routes.tokenRevocationURL, body);
-        return data;
+        await axios.post<void>(OAuth2Routes.tokenRevocationURL, body);
```
Comment (NachoToast) [inline src/classes/AuthDiscordAPI/AuthDiscordAPI.ts:62]: Turns out Discord's response to OAuth2 token revocation's isn't consistent.

Comment (NachoToast) [inline src/classes/AuthDiscordAPI/AuthDiscordAPI.ts:62]: Turns out Discord's response to OAuth2 token revocation isn't consistent, the return type here was actually wrong.

File: src/handlers/core/apply.ts [hunk 2]
```diff
@@ -73,42 +74,26 @@ export const apply: RequestHandler = async (req, res) => {
         });
     }
 
-    const guild = inviteValidation.invite.guild;
+    // make sure we dont have this guild in any of our databases already
+    {
+        for (const _state in EntriesDatabases) {
+            const state = _state as unknown as EntryStates;
```
Comment (tukib) [inline src/handlers/core/apply.ts:80]: purpose of the chained type assertion?

Comment (NachoToast) [reply src/handlers/core/apply.ts:80]: Keys of `EntriesDatabases` are `EntryStates`, however Typescript doesn't realize this.

File: src/handlers/entryModification/modifyEntryState.ts [hunk 3]
```diff
@@ -80,6 +75,15 @@ export const modifyEntryState: RequestHandler = (req, res) => {
         });
     }
 
+    // validate permissions for featured entries
+    if (currentState === EntryStates.Featured || newState === EntryStates.Featured) {
+        if (token.user.permissionLevel < UserPermissionLevels.Owner) {
+            res.setHeader(`Perms-RequiredLevel`, UserPermissionLevels.Owner);
+            res.setHeader(`Perms-CurrentLevel`, token.user.permissionLevel);
+            return res.sendStatus(401);
```
Comment (tukib) [inline src/handlers/entryModification/modifyEntryState.ts:83]: I assume here the user has credentials? This should be a 403 (forbidden) status if credentials are provided but aren't sufficient.
```suggestion
            return res.sendStatus(403);
```

File: src/handlers/core/apply.ts (hunk 2 above)
Comment (NachoToast) [inline src/handlers/core/apply.ts:80]: Should really be:
```suggestion
            const state = Number(_state) as EntryStates;
```

Comment (tukib) [reply src/handlers/core/apply.ts]: ah, wacky

**Review by tukib — APPROVED** [review]
Looks good to me but consider the suggestion below

File: src/handlers/entryModification/modifyEntryState.ts (hunk 3 above)
Comment (NachoToast) [reply src/handlers/entryModification/modifyEntryState.ts:83]: Yeah we'd have to change all our endpoints for that (and the API docs), maybe another PR.

**Pull request merged**